### PR TITLE
pprofui: add support for collecting goroutines with labels

### PR DESCRIPTION
**Pull request by adityamaru**
This change refactors some of the logic in the pprofui/. Additionally, we add support for collecting cluster-wide goroutine profiles.

These goroutine profiles can be collected with or without labels. When collected without labels we request pprof profiles with debug=0, which generates Profile protobufs on every node and `pprof.Merge`s them before redirecting the client to a flamegraph.

When collected with labels we request pprof profiles with debug=1, which generates a legacy text format with comments translating addresses to function names and line numbers, so that a programmer can read the profile without tools. These profiles cannot be `pprof.Merged` and so we manually stitch them together on a per-node basis before downloading them for the client as txt file.

This change also adds support for filtering the aforementioned goroutines by pprof label. This will be used by the Jobs Profiler to collect cluster-wide stacks relevant to the running job.

Informs: #105440
Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Tested on `backup/2TB/n10cpu4` and collected a cluster-wide stack:

`http://adityamaru-1688173202-01-n10cpu4-0001.roachprod.crdb.io:26258/debug/pprof/ui/goroutine/?node=all&labelfilter=IMPORT`

[goroutine_6.txt](https://github.com/cockroachdb/cockroach/files/11923907/goroutine_6.txt)

File: pkg/cli/zip_per_node.go
Comment (tbg) [inline pkg/cli/zip_per_node.go:110]: for future PRs: please put trivial refactors like this in separate commits.

Comment (tbg) [reply pkg/cli/zip_per_node.go:110]: Also, is this a backwards-incompatible change? I believe profiles can be requested from this endpoint via JSON, so this would break. Let's avoid unnecessary renames unless they are confirmed as benign by @cockroachdb/obs-inf-prs.

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:100]:
```suggestion
// validateProfileRequest validates that the request does not contain any
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:107]:
```suggestion
			return errors.Newf("filtering using a pprof label is unsupported for %s", req.Type)
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:111]:
```suggestion
			return errors.Newf("cluster-wide collection is unsupported for %s", req.Type)
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:115]:
```suggestion
			return errors.Newf("profiling with labels is unsupported for %s", req.Type)
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:119]:
```suggestion
			return errors.Newf("filtering using a pprof label is unsupported for %s", req.Type)
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:126]: Is this code movement? For future PRs, see if there is a way to have a separate commit that mechanically moves code (and the reviewer can trust that's all that's happening) and then layer additional modifications of the logic on top in separate commits. As is, I basically have to review from scratch, which is a lot more work.

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:139]: It's nice that they made that rename upstream. TIL.

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:199]: Glad to see that my legacy lives on 🙈

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:305]: I've seen this before, why do we have this twice?

File: pkg/server/status.go
Comment (tbg) [inline pkg/server/status.go:1470]: Is this useful? As is, the caller won't know the information is incomplete, so the UX for making this discoverable is poor. Consider adding the same plumbing for errors back into the response here. I see that the caller of this method gets to return a `JSONResponse`, so it would be straightforward to return an additional slice `[]error` from this method and put it into the `JSONResponse{Errors: errs.Map(func(err error) string { return err.Error() })}` (this doesn't exist but you get the idea).

File: pkg/server/status.go
Comment (tbg) [inline pkg/server/status.go:1475]: Ditto

File: pkg/server/status.go
Comment (tbg) [inline pkg/server/status.go:1490]: ditto

File: pkg/server/status.go
Comment (tbg) [inline pkg/server/status.go:1495]: ditto

File: pkg/server/status.go
Comment (tbg) [inline pkg/server/status.go:1580]:
```suggestion
		return nil, errors.Newf("cluster-wide collection of %s is unsupported", req.Type)
```

I fully expect you'll come back to me saying that the `.String` is necessary for some reason, but it shouldn't be. Perhaps `String` needs to be implemented on the pointer instead of value receiver or vice versa, either way we shouldn't have to call it explicitly here.

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:65]: 👑

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:68]: Interesting choice of `InvalidArgument`, I would just stick to internal error for all of these.

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:71]: totally up to you, just wanted to point out that `var buf bytes.NewBuffer` works too.

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:73]: nit: why the local `debug` and not just `req.WithLabels` everywhere?

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:123]: This is tested from `pprofui`, which seems right. So why not move it there, too? Is this the reason why you switched the pprofui tests to `pprofui_test`? If that's the case definitely move it and consider a switch back to `pprofui` for the test file.

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [inline pkg/server/status_local_file_retrieval.go:136]: Move this out to an init-type var. I know `regexp` has an internal cache but still!

**Review by tbg — CHANGES_REQUESTED** [review]
Looks good! I went a little nitty with the review just because I'm noticing that some practices we have "dialed in" on the replication team could be useful here too and I wanted to proselytise a little. Concretely, this could've been split up into a few commits where the logic changes are isolated and much easier to review. (As is, the review is a tad shallow, though I did make an effort at "really" reviewing the code). Please don't change it now, that's just busy work. I see this as an investment in the future - small commit style can be annoying at first, but in my opinion it's really worth it in that reviewers can much more easily engage with changes at a deeper level.

----

This is a great improvement, thanks for putting in the work! The main reason I'm requesting changes is that I feel strongly that non-fatal errors (missing nodes, etc) should be propagated back with the `JSONResponse`. I often look at clusters where we have partial data only (recently had one where it took >20 seconds just to render a 404...), and not knowing when you do can be difficult to deal with.

Other than that, mostly nits. If you haven't already, please give all of the profiles a smoke test. With changes like these, I can never quite be convinced that they work from code review alone, no matter how hard I look.

File: pkg/server/debug/pprofui/server.go
Comment (adityamaru) [reply pkg/server/debug/pprofui/server.go:126]: This was just code movement. Yeah I realize that a separate refactor only commit would have made this a lot easier. The truth is I didn't realize this change would balloon into what it did but I should've made the call to switch to smaller commits before getting in so deep. Noted for next time!

File: pkg/server/status_local_file_retrieval.go
Comment (adityamaru) [reply pkg/server/status_local_file_retrieval.go:123]: good point, moved to pprofui and moved the test file back to the original package.

File: pkg/cli/zip_per_node.go
Comment (adityamaru) [reply pkg/cli/zip_per_node.go:110]: noted, undid this rename

File: pkg/server/debug/pprofui/server.go
Comment (adityamaru) [reply pkg/server/debug/pprofui/server.go:305]: This logic exists to serve a newly generated profile. The same logic also exists to serve a cached profile request by profile ID.

File: pkg/server/status.go
Comment (adityamaru) [reply pkg/server/status.go:1470]: This makes sense but I'm not sure how we'd surface these `Errors` when writing the data to the http.ResponseWriter. In the case of all profiles except goroutine profiles with labels we redirect the user to a flamegraph with the serialized profile data in the response body. I was thinking of using the `Warning` response header - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Warning but this is marked as deprecated.

We could do the following:
- Set the `Comments` field in the `Profile` to contain the partial errors so that when the user switches to the `Peek` view they can notice the missing nodes. This would cover cpu profiles and goroutine profiles without labels.
- When goroutine profiles are set with labels we can log the partial errors at the top of the file that is queued to download.

Let me know if this is not what you had in mind!

File: pkg/server/status.go
Comment (adityamaru) [reply pkg/server/status.go]: I don't think I had a good reason to add .String suffixes. Dropped them and the test that checks for this type of error during request validation still passes.

File: pkg/server/status_local_file_retrieval.go
Comment (tbg) [reply pkg/server/status_local_file_retrieval.go:73]: My thought was to `s/debug/req.WithLabels/g`. There's just one use of the `debug` local. Easier to read, for me at least, if the indirection is avoided. This is minor, so your call!

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:343]:
```suggestion
// any one of the expectedLabels to resBuf.
// The input must be in the format used by debug/pprof/goroutines?debug=1.
```

File: pkg/server/debug/pprofui/server.go
Comment (tbg) [inline pkg/server/debug/pprofui/server.go:353]: We've had some problems[^1] with graphviz breaking when label values contained non-alphanumerics, do you think this will be an issue here? Concretely, `"` and `{` `}`. I wonder if it could be more robust to parse what follows `labels: ` via `json.Decode`, but Hopefully `"` and the like are escaped properly.

[^1]: https://github.com/cockroachdb/cockroach/issues/101523

**Review by tbg — APPROVED** [review]
💯 looking good, thanks again! Looking at the label parsing again, it seems perhaps a little brittle, would appreciate if you could look that over once more and maybe add a more adversarial test using funky label values.

CI complains about `./dev gen`.

File: pkg/server/debug/pprofui/server.go
Comment (adityamaru) [reply pkg/server/debug/pprofui/server.go:355]: Interesting, thanks for calling it out. I added some tests to ensure we handle `"` and `{` and `}`. The regexp seems to correctly match against these non-alphanumeric labels too. Considering goroutines with `debug=1` will only be available in txt format I don't think such labels break any of our logic.

**Comment by adityamaru** [comment]
TFTR!

bors r=tbg

**Pull request merged**
